Documentation Doxygéne/man/man3/video.3
.TH "video" 3 "Wed Sep 11 2013" "MARCS" \" -*- nroff -*-
.ad l
.nh
.SH NAME
video \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include 'video\&.h'\fP
.PP
Inherits QMainWindow\&.
.SS "Public Slots"
.SS "Public Member Functions"
.SS "Private Attributes"
.SH "Detailed Description"
.PP 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "\fBvideo\fP (QWidget *parent = \fC0\fP)\fC [explicit]\fP"

.PP
.nf
65                             :
66     QMainWindow(parent),
67     ui(new Ui::video)
68 {
69     affichageList = false ;//etat d'affichage de la liste du log
70     ui->setupUi(this);
71     place =  new GeoDataPlacemark( "" );
72     document = new GeoDataDocument();
73 
74 
75     //config marble
76     ui->MarbleWidget_smallView->setShowOverviewMap(false);
77     ui->MarbleWidget_plan->setShowOverviewMap(false);
78     ui->MarbleWidget_smallView->setShowScaleBar(false);
79     ui->MarbleWidget_smallView->setShowCompass(false);
80     ui->MarbleWidget_smallView->hide();
81 
82 
83     //initialisation de l'intreface
84     ui->tableWidget->hide();
85     ui->snapShoot_button->hide();
86     ui->actionFlight_plan->setEnabled(false);
87     ui->actionFlight_data->setEnabled(true);
88     ui->actionVideo->setEnabled(true);
89     ui->listLog->hide();
90     ui->actionClear_mission->setEnabled(false);
91     ui->label_Alt->hide();
92     ui->label_HDG->hide();
93     ui->label_Num_Waypoint->hide();
94     ui->label_Time->hide();
95     ui->lineEdit_HDG->hide();
96     ui->lineEdit_Alt->hide();
97     ui->timeEdit_Mission->hide();
98     ui->line_2->hide();
99     ui->SaveEdit_button->hide();
100     ui->LoadData_button->hide();
101     ui->comboBox_ListWaypoint->hide();
102     ui->FinishEdit_button_2->hide();
103 
104 
105 
106     //signal&&slot de l'application
107     connect(ui->actionFlight_plan,SIGNAL(triggered()),this,SLOT(openNewWindowMain()));
108     connect(ui->actionFlight_data,SIGNAL(triggered()),this,SLOT(openNewWindowData()));
109     connect(ui->actionVideo,SIGNAL(triggered()),this,SLOT(openNewWindowVideo()));
110     connect(ui->actionLoad_map,SIGNAL(triggered()),this,SLOT(openNewMap()));
111     connect(ui->actionLoad_mission,SIGNAL(triggered()),this,SLOT(openMission()));
112     connect(ui->actionClose,SIGNAL(triggered()),this,SLOT(close()));
113     connect(ui->MarbleWidget_smallView,SIGNAL(mouseClickGeoPosition(qreal,qreal,GeoDataCoordinates::Unit)),this,SLOT(switchToMap()));
114     connect(ui->status_button,SIGNAL(clicked()),this,SLOT(afficheList()));
115     connect(ui->led_button,SIGNAL(clicked()),this,SLOT(afficheList()));
116     connect(ui->actionSave_mission,SIGNAL(triggered()),this,SLOT(saveMission()));
117     connect(ui->actionClear_mission,SIGNAL(triggered()),this,SLOT(clearMission()));
118     connect(ui->actionStart_planning,SIGNAL(triggered()),this,SLOT(activateAddingPoint()));
119     connect(ui->actionEdit_waypoint,SIGNAL(triggered()),this,SLOT(editWaypoint()));
120 
121     QPaintDevice *paintDevice = this;
122    // QPaintEvent *evt ;
123     QImage image;
124     if (!isEnabled())
125     {
126         // If the globe covers fully the screen then we can use the faster
127         // RGB32 as there are no translucent areas involved\&.
128         QImage::Format imageFormat = ( ui->MarbleWidget_plan->viewport()->mapCoversViewport() )
129                                      ? QImage::Format_RGB32
130                                      : QImage::Format_ARGB32_Premultiplied;
131         // Paint to an intermediate image
132         image = QImage( rect()\&.size(), imageFormat );
133 
134         image\&.fill( Qt::transparent );
135         paintDevice = &image;
136 
137     }
138 
139 
140 gp = new GeoPainter ( paintDevice, ui->MarbleWidget_plan->viewport(), ui->MarbleWidget_plan->mapQuality() );
141 //ui->MarbleWidget_plan->paint( gp, evt->rect() );
142 drawMission();
143 
144 }
.fi
.SS "~\fBvideo\fP ()"

.PP
.nf
147 {
148     delete ui;
149 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "void activateAddingPoint ()\fC [slot]\fP"

.PP
.nf
162                                {
163     ui->actionSave_mission->setEnabled(true);
164     connect(ui->MarbleWidget_plan,SIGNAL(mouseClickGeoPosition(qreal,qreal,GeoDataCoordinates::Unit)),this,SLOT(addPoint(qreal,qreal,GeoDataCoordinates::Unit)));
165     ui->actionStart_planning->setEnabled(false);
166 }
.fi
.SS "void addPoint (qreallon, qreallat, GeoDataCoordinates::Unit)\fC [slot]\fP"

.PP
.nf
268                                                                 {
269 
270 
271   sprintf(numWpText,"%d",num_waypoint);
272   temp = string(numWpText);
273   textNumWaypoint = "# " + temp ;
274   qstr = QString::fromStdString(textNumWaypoint);
275 
276   // Access the shared route request (start, destination and parameters)
277   manager = ui->MarbleWidget_plan->model()->routingManager();
278   manager_smallMap = ui->MarbleWidget_smallView->model()->routingManager();
279   request = manager->routeRequest();
280   request_smallMap = manager_smallMap->routeRequest();
281 
282   tempo = new GeoDataCoordinates(lon,lat,0\&.0,GeoDataCoordinates::Radian);
283 
284  // add point to map
285  request->append( GeoDataCoordinates( lon, lat, 0\&.0, GeoDataCoordinates::Radian ) );
286  request->setName(num_waypoint,qstr);
287  request_smallMap->append(GeoDataCoordinates( lon, lat, 0\&.0, GeoDataCoordinates::Radian ));
288  request_smallMap->setName(num_waypoint,qstr);
289  wpListSave\&.append(new waypoint(num_waypoint,tempo->longitude(GeoDataCoordinates::Degree),tempo->latitude(GeoDataCoordinates::Degree),100\&.0,90\&.0,60,1,1));
290  num_waypoint++ ;
291 
292  ui->actionEdit_waypoint->setEnabled(true);
293  open = false ;
294 }
.fi
.SS "void afficheList ()\fC [slot]\fP"

.PP
.nf
170 {
171     if (affichageList==false){
172         ui->listLog->show();
173         affichageList = true ;
174     }
175     else {
176         ui->listLog->hide();
177         affichageList = false ;
178     }
179 
180 }
.fi
.SS "void clearMission ()\fC [slot]\fP"

.PP
.nf
153                         {
154 
155     ui->MarbleWidget_plan->model()->removeGeoData(lastMission);
156     ui->actionClear_mission->setEnabled(false);
157     ui->actionSave_mission->setEnabled(false);
158 
159 }
.fi
.SS "void close ()\fC [slot]\fP"

.PP
.nf
340 {
341     exit(0);
342 }
.fi
.SS "void drawMission ()\fC [slot]\fP"

.PP
.nf
372                        {
373 
374     GeoDataCoordinates France( 2\&.2, 48\&.52, 0\&.0, GeoDataCoordinates::Degree );
375     gp->setPen( QColor( 0, 0, 0 ) );
376     gp->drawText( France, "France" );
377 
378     GeoDataCoordinates Canada( -77\&.02, 48\&.52, 0\&.0, GeoDataCoordinates::Degree );
379     gp->setPen( QColor( 0, 0, 0 ) );
380     gp->drawText( Canada, "Canada" );
381 
382     //A line from France to Canada without tessellation
383 
384     GeoDataLineString shapeNoTessellation( NoTessellation );
385     shapeNoTessellation << France << Canada;
386 
387     gp->setPen( oxygenSkyBlue4 );
388     gp->drawPolyline( shapeNoTessellation );
389 
390     //The same line, but with tessellation
391 
392     GeoDataLineString shapeTessellate( Tessellate );
393     shapeTessellate << France << Canada;
394 
395     gp->setPen( oxygenBrickRed4 );
396     gp->drawPolyline( shapeTessellate );
397 
398     //Now following the latitude circles
399 
400     GeoDataLineString shapeLatitudeCircle( RespectLatitudeCircle | Tessellate );
401     shapeLatitudeCircle << France << Canada;
402 
403     gp->setPen( oxygenForestGreen4 );
404     gp->drawPolyline( shapeLatitudeCircle );
405 }
.fi
.SS "void editWaypoint ()\fC [slot]\fP"

.PP
.nf
407                         {
408     ui->comboBox_ListWaypoint->show();
409     ui->LoadData_button->show();
410     ui->label_Num_Waypoint->show();
411 
412     ui->actionEdit_waypoint->setEnabled(false);
413     ui->comboBox_ListWaypoint->clear();
414 
415     if ( open == false){
416 
417         for ( int i = 0 ; i < wpListSave\&.size(); i++){
418             ui->comboBox_ListWaypoint->addItem(QString::number(i+1));
419         }
420     }
421     else {
422         for ( int i = 0 ; i < wpListOpen\&.size(); i++){
423             ui->comboBox_ListWaypoint->addItem(QString::number(i+1));
424         }
425 
426     }
427 
428     connect(ui->LoadData_button,SIGNAL(clicked()),this,SLOT(loadData()));
429 
430 
431 }
.fi
.SS "void finishEditData ()\fC [slot]\fP"

.PP
.nf
504                           {
505 
506 
507     ui->label_Alt->hide();
508     ui->label_HDG->hide();
509     ui->label_Num_Waypoint->hide();
510     ui->label_Time->hide();
511     ui->lineEdit_HDG->hide();
512     ui->lineEdit_Alt->hide();
513     ui->timeEdit_Mission->hide();
514     ui->line_2->hide();
515     ui->SaveEdit_button->hide();
516     ui->LoadData_button->hide();
517     ui->comboBox_ListWaypoint->hide();
518     ui->FinishEdit_button_2->hide();
519     ui->actionEdit_waypoint->setEnabled(true);
520 
521 
522 }
.fi
.SS "void loadData ()\fC [slot]\fP"

.PP
.nf
433                     {
434     int h ;
435     int m ;
436     int s ;
437 
438     ui->label_Alt->show();
439     ui->label_HDG->show();
440     ui->label_Time->show();
441     ui->lineEdit_Alt->show();
442     ui->lineEdit_HDG->show();
443     ui->timeEdit_Mission->show();
444     ui->line_2->show();
445     ui->SaveEdit_button->show();
446     ui->FinishEdit_button_2->show();
447     ui->SaveEdit_button->setEnabled(true);
448 
449     if ( open == false){
450         h = wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->getTime() / 3600 ;
451         m = (wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->getTime() % 3600)/60 ;
452         s = (wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->getTime() % 3600)%60  ;
453 
454        ui->lineEdit_Alt->setText(QString::number(wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->getAlt()));
455        ui->lineEdit_HDG->setText(QString::number(wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->getHdg()));
456        ui->timeEdit_Mission->setTime(QTime(h,m,s,00));
457 
458     }
459 
460     else {
461         h = wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->getTime() / 3600 ;
462         m = (wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->getTime() % 3600)/60 ;
463         s = (wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->getTime() % 3600)%60  ;
464 
465        ui->lineEdit_Alt->setText(QString::number(wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->getAlt()));
466        ui->lineEdit_HDG->setText(QString::number(wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->getHdg()));
467        ui->timeEdit_Mission->setTime(QTime(h,m,s,00));
468 
469     }
470 
471     connect(ui->SaveEdit_button,SIGNAL(clicked()),this,SLOT(saveEditData()));
472     connect(ui->FinishEdit_button_2,SIGNAL(clicked()),this,SLOT(finishEditData()));
473 
474    }
.fi
.SS "void openMission ()\fC [slot]\fP"

.PP
.nf
183                        {
184 
185    disconnect(ui->MarbleWidget_plan,SIGNAL(mouseClickGeoPosition(qreal,qreal,GeoDataCoordinates::Unit)),this,SLOT(addPoint(qreal,qreal,GeoDataCoordinates::Unit)));
186 
187     path = QFileDialog::getOpenFileName(this, "Load a mission", "C:/Users/mboussai/Desktop/Mohamed/Qt Projects/FILES/mission", "KML (*\&.kml)");
188     QFileInfo inputFile(path);
189     QString file ;
190     file = "C:/Users/mboussai/Desktop/Mohamed/Qt Projects/MARCS/MissionXML/testParse\&.xml";
191 
192     string m ;
193     string nameFileOpen;
194 
195 
196     m = path\&.toUtf8()\&.constData();
197     size_t pos = m\&.find_last_of("/");
198     if(pos != std::string::npos){
199 
200         nameFileOpen\&.assign(m\&.begin() + pos + 1, m\&.end()-4);
201     }
202       else
203     {
204       nameFileOpen = m;
205   }
206     string fileNameTempOpen = "C:/Users/mboussai/Desktop/Mohamed/Qt Projects/MARCS/MissionXML/" + nameFileOpen +"\&.xml" ;
207     fileOpened =  QString::fromStdString(fileNameTempOpen);
208 
209     // Access the shared route request (start, destination and parameters)
210     manager = ui->MarbleWidget_plan->model()->routingManager();
211     manager_smallMap = ui->MarbleWidget_smallView->model()->routingManager();
212     request = manager->routeRequest();
213     request_smallMap = manager_smallMap->routeRequest();
214     request->append( GeoDataCoordinates( 0\&.0, 1\&.0, 0\&.0, GeoDataCoordinates::Radian ) );
215     request_smallMap->append(GeoDataCoordinates( 0\&.0, 4\&.0, 0\&.0, GeoDataCoordinates::Radian ));
216 
217     request->clear();
218     request_smallMap->clear();
219     ui->MarbleWidget_plan->model()->addGeoDataFile( inputFile\&.absoluteFilePath() );
220     ui->MarbleWidget_plan->repaint();
221     ui->MarbleWidget_smallView->model()->addGeoDataFile( inputFile\&.absoluteFilePath() );
222     lastMission = inputFile\&.absoluteFilePath();
223     ui->actionClear_mission->setEnabled(true);
224     lastMission = inputFile\&.absoluteFilePath();
225     ui->actionStart_planning->setEnabled(true);
226 
227      wpListOpen = myMission\&.loadMission(wpListOpen,fileOpened);
228     ui->actionEdit_waypoint->setEnabled(true);
229     open = true ;
230 
231 
232 }
.fi
.SS "void openNewMap ()\fC [slot]\fP"

.PP
.nf
347 {
348     path = QFileDialog::getOpenFileName(this, "Load a map","C:/Users/mboussai/Desktop/Mohamed/Qt Projects/FILES/map", "OSM (*\&.osm)");
349     QFileInfo inputFile(path);
350 
351         ui->MarbleWidget_plan->model()->removeGeoData(lastMap);
352         ui->MarbleWidget_plan->model()->addGeoDataFile( inputFile\&.absoluteFilePath() );

MARCS/video.cpp
#include "video.h"
#include "ui_video.h"
#include "mission.h"
#include "waypoint.h"
#include <global.h>
#include <AbstractFloatItem.h>
#include <QPaintDevice>
#include <QString>
#include <QtGlobal>
#include <QFileDialog>
#include <QtCore/QFileInfo>
#include <QtGui/QApplication>
#include <QListView>
#include <string>
#include <iostream>
#include <QtGui/QTreeView>
#include <QtCore/QDataStream>
#include <QtCore/QFile>
#include <QtCore/QIODevice>
#include <QListWidget>
#include <QListWidgetItem>
#include <QtGui/QTreeView>
#include <MarbleWidget.h>
#include <MarbleModel.h>
#include <MarbleMap.h>
#include <GeoPainter.h>
#include <RouteRequest.h>
#include <RoutingManager.h>
#include <GeoDataDocument.h>
#include <GeoDataPlacemark.h>
#include <GeoDataLineString.h>
#include <GeoDataTreeModel.h>
#include <RoutingWidget.h>
#include <GeoPainter.h>
#include <ViewportParams.h>
#include <cstdio>
#include <sstream>

using namespace std;
using namespace Marble;

QListView *m_listView;
bool affichageList;
GeoDataPlacemark *place ;
GeoDataDocument *document ;
GeoDataCoordinates *tempo;
RoutingManager* manager ;
RouteRequest* request;
RoutingManager* manager_smallMap ;
RouteRequest* request_smallMap;
QList < waypoint* > wpListSave ;
QList <waypoint* > wpListOpen ;
QString lastMission;
QString lastMap;
int num_waypoint = 0;
string textNumWaypoint ;
string temp;
QString qstr ;
bool open = false ;
char* numWpText = new char[32];
GeoPainter* gp ;
QString fileOpened ;


video::video(QWidget *parent) :
    QMainWindow(parent),
    ui(new Ui::video)
{
    affichageList = false ;//etat d'affichage de la liste du log
    ui->setupUi(this);
    place =  new GeoDataPlacemark( "" );
    document = new GeoDataDocument();


    //config marble
    ui->MarbleWidget_smallView->setShowOverviewMap(false);
    ui->MarbleWidget_plan->setShowOverviewMap(false);
    ui->MarbleWidget_smallView->setShowScaleBar(false);
    ui->MarbleWidget_smallView->setShowCompass(false);
    ui->MarbleWidget_smallView->hide();


    //initialisation de l'intreface
    ui->tableWidget->hide();
    ui->snapShoot_button->hide();
    ui->actionFlight_plan->setEnabled(false);
    ui->actionFlight_data->setEnabled(true);
    ui->actionVideo->setEnabled(true);
    ui->listLog->hide();
    ui->actionClear_mission->setEnabled(false);
    ui->label_Alt->hide();
    ui->label_HDG->hide();
    ui->label_Num_Waypoint->hide();
    ui->label_Time->hide();
    ui->lineEdit_HDG->hide();
    ui->lineEdit_Alt->hide();
    ui->timeEdit_Mission->hide();
    ui->line_2->hide();
    ui->SaveEdit_button->hide();
    ui->LoadData_button->hide();
    ui->comboBox_ListWaypoint->hide();
    ui->FinishEdit_button_2->hide();



    //signal&&slot de l'application
    connect(ui->actionFlight_plan,SIGNAL(triggered()),this,SLOT(openNewWindowMain()));
    connect(ui->actionFlight_data,SIGNAL(triggered()),this,SLOT(openNewWindowData()));
    connect(ui->actionVideo,SIGNAL(triggered()),this,SLOT(openNewWindowVideo()));
    connect(ui->actionLoad_map,SIGNAL(triggered()),this,SLOT(openNewMap()));
    connect(ui->actionLoad_mission,SIGNAL(triggered()),this,SLOT(openMission()));
    connect(ui->actionClose,SIGNAL(triggered()),this,SLOT(close()));
    connect(ui->MarbleWidget_smallView,SIGNAL(mouseClickGeoPosition(qreal,qreal,GeoDataCoordinates::Unit)),this,SLOT(switchToMap()));
    connect(ui->status_button,SIGNAL(clicked()),this,SLOT(afficheList()));
    connect(ui->led_button,SIGNAL(clicked()),this,SLOT(afficheList()));
    connect(ui->actionSave_mission,SIGNAL(triggered()),this,SLOT(saveMission()));
    connect(ui->actionClear_mission,SIGNAL(triggered()),this,SLOT(clearMission()));
    connect(ui->actionStart_planning,SIGNAL(triggered()),this,SLOT(activateAddingPoint()));
    connect(ui->actionEdit_waypoint,SIGNAL(triggered()),this,SLOT(editWaypoint()));

    QPaintDevice *paintDevice = this;
   // QPaintEvent *evt ;
    QImage image;
    if (!isEnabled())
    {
        // If the globe covers fully the screen then we can use the faster
        // RGB32 as there are no translucent areas involved.
        QImage::Format imageFormat = ( ui->MarbleWidget_plan->viewport()->mapCoversViewport() )
                                     ? QImage::Format_RGB32
                                     : QImage::Format_ARGB32_Premultiplied;
        // Paint to an intermediate image
        image = QImage( rect().size(), imageFormat );

        image.fill( Qt::transparent );
        paintDevice = &image;

    }


gp = new GeoPainter ( paintDevice, ui->MarbleWidget_plan->viewport(), ui->MarbleWidget_plan->mapQuality() );
//ui->MarbleWidget_plan->paint( gp, evt->rect() );
drawMission();

}

video::~video()
{
    delete ui;
}


//Delete a mission from the view
void video::clearMission(){

    ui->MarbleWidget_plan->model()->removeGeoData(lastMission);
    ui->actionClear_mission->setEnabled(false);
    ui->actionSave_mission->setEnabled(false);

}

//Activate the adding points on the map
void video::activateAddingPoint(){
    ui->actionSave_mission->setEnabled(true);
    connect(ui->MarbleWidget_plan,SIGNAL(mouseClickGeoPosition(qreal,qreal,GeoDataCoordinates::Unit)),this,SLOT(addPoint(qreal,qreal,GeoDataCoordinates::Unit)));
    ui->actionStart_planning->setEnabled(false);
}

//Show the list of logs
void video::afficheList()
{
    if (affichageList==false){
        ui->listLog->show();
        affichageList = true ;
    }
    else {
        ui->listLog->hide();
        affichageList = false ;
    }

}
//open a mission

void video::openMission(){

   disconnect(ui->MarbleWidget_plan,SIGNAL(mouseClickGeoPosition(qreal,qreal,GeoDataCoordinates::Unit)),this,SLOT(addPoint(qreal,qreal,GeoDataCoordinates::Unit)));

    path = QFileDialog::getOpenFileName(this, "Load a mission", "C:/Users/mboussai/Desktop/Mohamed/Qt Projects/FILES/mission", "KML (*.kml)");
    QFileInfo inputFile(path);
    QString file ;
    file = "C:/Users/mboussai/Desktop/Mohamed/Qt Projects/MARCS/MissionXML/testParse.xml";

    string m ;
    string nameFileOpen;


    m = path.toUtf8().constData();
    size_t pos = m.find_last_of("/");
    if(pos != std::string::npos){

        nameFileOpen.assign(m.begin() + pos + 1, m.end()-4);
    }
      else
    {
      nameFileOpen = m;
  }
    string fileNameTempOpen = "C:/Users/mboussai/Desktop/Mohamed/Qt Projects/MARCS/MissionXML/" + nameFileOpen +".xml" ;
    fileOpened =  QString::fromStdString(fileNameTempOpen);

    // Access the shared route request (start, destination and parameters)
    manager = ui->MarbleWidget_plan->model()->routingManager();
    manager_smallMap = ui->MarbleWidget_smallView->model()->routingManager();
    request = manager->routeRequest();
    request_smallMap = manager_smallMap->routeRequest();
    request->append( GeoDataCoordinates( 0.0, 1.0, 0.0, GeoDataCoordinates::Radian ) );
    request_smallMap->append(GeoDataCoordinates( 0.0, 4.0, 0.0, GeoDataCoordinates::Radian ));

    request->clear();
    request_smallMap->clear();
    ui->MarbleWidget_plan->model()->addGeoDataFile( inputFile.absoluteFilePath() );
    ui->MarbleWidget_plan->repaint();
    ui->MarbleWidget_smallView->model()->addGeoDataFile( inputFile.absoluteFilePath() );
    lastMission = inputFile.absoluteFilePath();
    ui->actionClear_mission->setEnabled(true);
    lastMission = inputFile.absoluteFilePath();
    ui->actionStart_planning->setEnabled(true);

     wpListOpen = myMission.loadMission(wpListOpen,fileOpened);
    ui->actionEdit_waypoint->setEnabled(true);
    open = true ;


}
//Save a mission

void video::saveMission(){
    QString  fileName = QFileDialog::getSaveFileName( this,  "Save a mission" , "C:/Users/mboussai/Desktop/Mohamed/Qt Projects/FILES/mission",  "KML files (*.kml)"  );
    string m ;
    string nameFile;
    QString hideFile ;


    m = fileName.toUtf8().constData();
    size_t pos = m.find_last_of("/");
    if(pos != std::string::npos){

        nameFile.assign(m.begin() + pos + 1, m.end()-4);
    }
      else
    {
      nameFile = m;
  }
    string fileNameTemp = "C:/Users/mboussai/Desktop/Mohamed/Qt Projects/MARCS/MissionXML/" + nameFile +".xml" ;
    hideFile =  QString::fromStdString(fileNameTemp);

      if ( !fileName.isEmpty() ) {
          manager->saveRoute(fileName);

       }

    myMission.saveMission(wpListSave,hideFile);

    ui->actionSave_mission->setEnabled(false);

}


//add waypoint to the Flight Plan
void video::addPoint(qreal lon, qreal lat, GeoDataCoordinates::Unit){


  sprintf(numWpText,"%d",num_waypoint);
  temp = string(numWpText);
  textNumWaypoint = "# " + temp ;
  qstr = QString::fromStdString(textNumWaypoint);

  // Access the shared route request (start, destination and parameters)
  manager = ui->MarbleWidget_plan->model()->routingManager();
  manager_smallMap = ui->MarbleWidget_smallView->model()->routingManager();
  request = manager->routeRequest();
  request_smallMap = manager_smallMap->routeRequest();

  tempo = new GeoDataCoordinates(lon,lat,0.0,GeoDataCoordinates::Radian);

 // add point to map
 request->append( GeoDataCoordinates( lon, lat, 0.0, GeoDataCoordinates::Radian ) );
 request->setName(num_waypoint,qstr);
 request_smallMap->append(GeoDataCoordinates( lon, lat, 0.0, GeoDataCoordinates::Radian ));
 request_smallMap->setName(num_waypoint,qstr);
 wpListSave.append(new waypoint(num_waypoint,tempo->longitude(GeoDataCoordinates::Degree),tempo->latitude(GeoDataCoordinates::Degree),100.0,90.0,60,1,1));
 num_waypoint++ ;

 ui->actionEdit_waypoint->setEnabled(true);
 open = false ;
}

//Show the page "Flight Plan"
void video::openNewWindowMain()
{
    ui->MarbleWidget_plan->show();
    ui->MarbleWidget_smallView->hide();
    ui->tableWidget->hide();
    ui->snapShoot_button->hide();
    ui->addMark_button->show();
    ui->actionFlight_plan->setEnabled(false);
    ui->actionFlight_data->setEnabled(true);
    ui->actionVideo->setEnabled(true);

}

//Show the page "Flight DATA"
void video::openNewWindowData()
{
    ui->MarbleWidget_plan->hide();
    ui->MarbleWidget_smallView->hide();
    ui->tableWidget->show();
    ui->snapShoot_button->hide();
    ui->addMark_button->hide();
    ui->actionFlight_plan->setEnabled(true);
    ui->actionFlight_data->setEnabled(false);
    ui->actionVideo->setEnabled(true);

}

//Show the page "VIDEO"
void video::openNewWindowVideo()
{
    ui->MarbleWidget_plan->hide();
    ui->MarbleWidget_smallView->show();
    ui->tableWidget->hide();
    ui->snapShoot_button->show();
    ui->addMark_button->show();
    ui->actionFlight_plan->setEnabled(true);
    ui->actionFlight_data->setEnabled(true);
    ui->actionVideo->setEnabled(false);

}

//Close the window
void video::close()
{
    exit(0);
}

//Load a map with the format .osm
void video::openNewMap()

{
    path = QFileDialog::getOpenFileName(this, "Load a map","C:/Users/mboussai/Desktop/Mohamed/Qt Projects/FILES/map", "OSM (*.osm)");
    QFileInfo inputFile(path);

        ui->MarbleWidget_plan->model()->removeGeoData(lastMap);
        ui->MarbleWidget_plan->model()->addGeoDataFile( inputFile.absoluteFilePath() );
        ui->MarbleWidget_smallView->model()->addGeoDataFile( inputFile.absoluteFilePath() );
        lastMap = inputFile.absoluteFilePath();

}
//Switch Mode ( Flight Plan , Video )

void video::switchToMap()
{

    ui->MarbleWidget_plan->show();
    ui->MarbleWidget_smallView->hide();
    ui->tableWidget->hide();
    ui->snapShoot_button->hide();
    ui->addMark_button->show();
    ui->actionFlight_plan->setEnabled(false);
    ui->actionFlight_data->setEnabled(true);
    ui->actionVideo->setEnabled(true);
}

void video::drawMission(){

    GeoDataCoordinates France( 2.2, 48.52, 0.0, GeoDataCoordinates::Degree );
    gp->setPen( QColor( 0, 0, 0 ) );
    gp->drawText( France, "France" );

    GeoDataCoordinates Canada( -77.02, 48.52, 0.0, GeoDataCoordinates::Degree );
    gp->setPen( QColor( 0, 0, 0 ) );
    gp->drawText( Canada, "Canada" );

    //A line from France to Canada without tessellation

    GeoDataLineString shapeNoTessellation( NoTessellation );
    shapeNoTessellation << France << Canada;

    gp->setPen( oxygenSkyBlue4 );
    gp->drawPolyline( shapeNoTessellation );

    //The same line, but with tessellation

    GeoDataLineString shapeTessellate( Tessellate );
    shapeTessellate << France << Canada;

    gp->setPen( oxygenBrickRed4 );
    gp->drawPolyline( shapeTessellate );

    //Now following the latitude circles

    GeoDataLineString shapeLatitudeCircle( RespectLatitudeCircle | Tessellate );
    shapeLatitudeCircle << France << Canada;

    gp->setPen( oxygenForestGreen4 );
    gp->drawPolyline( shapeLatitudeCircle );
}

void video::editWaypoint(){
    ui->comboBox_ListWaypoint->show();
    ui->LoadData_button->show();
    ui->label_Num_Waypoint->show();

    ui->actionEdit_waypoint->setEnabled(false);
    ui->comboBox_ListWaypoint->clear();

    if ( open == false){

        for ( int i = 0 ; i < wpListSave.size(); i++){
            ui->comboBox_ListWaypoint->addItem(QString::number(i+1));
        }
    }
    else {
        for ( int i = 0 ; i < wpListOpen.size(); i++){
            ui->comboBox_ListWaypoint->addItem(QString::number(i+1));
        }

    }

    connect(ui->LoadData_button,SIGNAL(clicked()),this,SLOT(loadData()));


}

void video::loadData(){
    int h ;
    int m ;
    int s ;

    ui->label_Alt->show();
    ui->label_HDG->show();
    ui->label_Time->show();
    ui->lineEdit_Alt->show();
    ui->lineEdit_HDG->show();
    ui->timeEdit_Mission->show();
    ui->line_2->show();
    ui->SaveEdit_button->show();
    ui->FinishEdit_button_2->show();
    ui->SaveEdit_button->setEnabled(true);

    if ( open == false){
        h = wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->getTime() / 3600 ;
        m = (wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->getTime() % 3600)/60 ;
        s = (wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->getTime() % 3600)%60  ;

       ui->lineEdit_Alt->setText(QString::number(wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->getAlt()));
       ui->lineEdit_HDG->setText(QString::number(wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->getHdg()));
       ui->timeEdit_Mission->setTime(QTime(h,m,s,00));

    }

    else {
        h = wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->getTime() / 3600 ;
        m = (wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->getTime() % 3600)/60 ;
        s = (wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->getTime() % 3600)%60  ;

       ui->lineEdit_Alt->setText(QString::number(wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->getAlt()));
       ui->lineEdit_HDG->setText(QString::number(wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->getHdg()));
       ui->timeEdit_Mission->setTime(QTime(h,m,s,00));

    }

    connect(ui->SaveEdit_button,SIGNAL(clicked()),this,SLOT(saveEditData()));
    connect(ui->FinishEdit_button_2,SIGNAL(clicked()),this,SLOT(finishEditData()));

   }

void video::saveEditData(){
    bool ok = false ;
    QTime tempo = ui->timeEdit_Mission->time();
    int time = tempo.hour() * 3600 + tempo.minute() * 60 + tempo.second() ;

    if ( open == false){

        wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->setAlt(ui->lineEdit_Alt->text().toDouble(&ok));
        wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->setHdg(ui->lineEdit_HDG->text().toDouble(&ok));
        wpListSave[ui->comboBox_ListWaypoint->currentIndex()]->setTime(time);

    }

    else {


        wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->setAlt(ui->lineEdit_Alt->text().toDouble(&ok));
        wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->setHdg(ui->lineEdit_HDG->text().toDouble(&ok));
        wpListOpen[ui->comboBox_ListWaypoint->currentIndex()]->setTime(time);
        myMission.saveMission(wpListOpen,fileOpened);

    }


    ui->SaveEdit_button->setEnabled(false);

}

void video::finishEditData(){


    ui->label_Alt->hide();
    ui->label_HDG->hide();
    ui->label_Num_Waypoint->hide();
    ui->label_Time->hide();
    ui->lineEdit_HDG->hide();
    ui->lineEdit_Alt->hide();
    ui->timeEdit_Mission->hide();
    ui->line_2->hide();
    ui->SaveEdit_button->hide();
    ui->LoadData_button->hide();
    ui->comboBox_ListWaypoint->hide();
    ui->FinishEdit_button_2->hide();
    ui->actionEdit_waypoint->setEnabled(true);


}


